src/action-providers.ts

```diff
@@ -21,9 +21,73 @@ class RefactoringActionProvider implements vscode.CodeActionProvider {
     const ast = t.parse(document.getText());
     const selection = createSelectionFromVSCode(range);
 
-    return this.refactorings
-      .filter(refactoring => this.canPerform(refactoring, ast, selection))
-      .map(refactoring => this.buildCodeActionFor(refactoring));
+    const applicableRefactorings: RefactoringWithActionProvider[] = [];
+    t.traverseAST(ast, {
+      enter: (path: t.NodePath<any>) => {
+        this.refactorings.forEach(
+          this.checkCanPerform(
+            path,
+            selection,
+            (
+              matchedPath: t.NodePath<any>,
+              refactoring: RefactoringWithActionProvider
+            ) => {
+              if (refactoring.actionProvider.canPerformRefactoringMutator) {
+                applicableRefactorings.push(
+                  refactoring.actionProvider.canPerformRefactoringMutator(
+                    matchedPath,
+                    refactoring
+                  )
+                );
+              } else {
+                applicableRefactorings.push(refactoring);
+              }
+            }
+          )
+        );
+        path.type;
+      }
+    });
+
+    return [
+      ...applicableRefactorings.map(refactoring =>
+        this.buildCodeActionFor(refactoring)
+      ),
+      ...this.refactorings
+        .filter(refactoring => this.canPerform(refactoring, ast, selection))
+        .map(refactoring => this.buildCodeActionFor(refactoring))
+    ];
+  }
+
+  private checkCanPerform(
+    path: any,
+    selection: any,
+    onCanPerform: (
+      matchedPath: t.NodePath<any>,
+      refactoring: RefactoringWithActionProvider
+    ) => void
+  ) {
+    return (refactoring: RefactoringWithActionProvider) => {
+      if (refactoring.actionProvider.canPerformVisitorFactory) {
+        const visitor: any = refactoring.actionProvider.canPerformVisitorFactory(
+          selection,
+          path => onCanPerform(path, refactoring),
+          refactoring
+        );
+
+        const node: t.Node = path.node;
+
+        try {
+          if (typeof visitor[node.type] === "function") {
+            visitor[node.type](path);
+          } else if (typeof visitor[node.type] === "object") {
```

Comment:
I wonder if we could standardize this so it streamlines the code. But overall, I get the idea 👍

Reply:
Unfortunately, visitors can either be an object or a function. It would be confusing for devs to not be able to use one of those even though babel's AST traversal function explicitly can do both. However, I think this can be improved by extracting a function that encapsulates both variants.